poc_iot_injector/src/server.rs

```diff
@@ -139,85 +142,89 @@ async fn submit_txns(
-                            if do_submission {
-                                tracing::info!("submitting txn: {:?}", txn_details.hash_b64_url);
-                                if let Err(err) = handle_txn_submission(
-                                    txn_details.clone(),
-                                    &mut shared_txn_service,
-                                )
-                                .await
-                                {
-                                    tracing::warn!("unable to submit_txn, {err:?}");
-                                } else {
-                                    tracing::info!("storing txn: {:?}", txn_details.hash_b64_url);
-                                    metrics::increment_counter!("poc_injector_receipt_count");
-                                    let _ = file_sink_write!(
-                                        "signed_poc_receipt_txn",
-                                        receipt_sender,
-                                        txn_details.txn
-                                    )
-                                    .await;
-                                }
-                            } else {
-                                tracing::info!("only storing txn: {:?}", txn_details.hash_b64_url);
-                                metrics::increment_counter!("poc_injector_receipt_count");
-                                let _ = file_sink_write!(
-                                    "signed_poc_receipt_txn",
-                                    receipt_sender,
-                                    txn_details.txn
-                                )
-                                .await;
-                            }
-                        }
-                        Err(err) => tracing::warn!("unable to construct txn_details, {err:?}"),
-                    },
+    let mut stream =
+        store.source_unordered(LOADER_WORKERS, stream::iter(file_list).map(Ok).boxed());
+
+    let mut set = JoinSet::new();
+
+    while let Some(msg) = stream.next().await {
+        match msg {
+            Err(err) => tracing::warn!("skipping entry in stream: {err:?}"),
+            Ok(buf) => {
+                let shared_key = keypair.clone();
+                let mut shared_txn_service = txn_service.clone();
+                let receipt_sender_clone = receipt_sender.clone();
+
+                tracing::info!("Spawning submission tasks...");
+                set.spawn(async move {
+                    let _ = process_submission(
+                        buf,
+                        shared_key,
+                        do_submission,
+                        &receipt_sender_clone,
+                        &mut shared_txn_service,
+                    )
+                    .await;
+                });
+
+                if set.len() > MAX_CONCURRENT_SUBMISSIONS {
+                    tracing::info!("Processing {MAX_CONCURRENT_SUBMISSIONS} submissions");
+                    set.join_next().await;
```

Comment:
so will this effectively hold within the body of this `while let` loop until the stream has exhausted, and keep processing (`join` ing) the next message every time the queue exceeds the maximum length until it drops back down again? then when the stream _is_ exhausted it'll iterate through the `while` loop below until all of the outstanding tasks have been completed?

Reply:
Yeah, that's the idea at least.